### ISSUE TITLE
Custom interactions in control styled as a native control

### ISSUE THREAD

**Issue by karlgroves** [WordPress/gutenberg#15319]
Custom interactions in control styled as a native control

-   **Severity:** Medium
-   **Affected Populations:**
    -   Low-Vision
    -   Motor Impaired
    -   Cognitively Impaired
-   **Platform(s):**
    -   All / Universal
-   **Components affected:**
    -   Block Panel

#### Issue description

In the Block panel, where the font size can be selected, users see what
appears to be a select element. However when keyboard users reach this
component, they cannot use standard keystrokes to change the value, such
as Up-Arrow or Down-Arrow, or Alt+Arrow to open the submenu. Instead,
users need to press Enter to view the submenu.

Keyboard users will expect this control to work the same as a native
select element. Speech recognition users often speak out keystrokes to
perform standard actions on native components, and these would not work
on custom components like these.

##### Issue Code

```
    <button type="button" aria-expanded="false" aria-label="Font size: Normal" class="components-button components-font-size-picker__selector is-button is-default is-large">Normal</button>

```

#### Remediation Guidance

Use a native `<select>` element in place of a custom button+submenu
setup.

##### Recommended Code

```
    <fieldset class="components-base-control__field">


        <legend class="components-base-control__label">Font Size</legend>


        <label for="font-size-picker">Choose basic size</label>


        <select id-"font-size-picker" class="attachment-filters">


            <option value="normal" >Normal</option>


            ...


        </select>


        <label for="custom-font-size">Custom font size</label>


        <input id="custom-font-size" class="components-range-control__number" type="number" value="">


        <button type="button" disabled="" class="components-button components-color-palette__clear is-button is-default is-small">Reset</button>


    </fieldset>
```

#### Relevant standards

-   [2.1.1 Keyboard
    (Level A)](https://www.w3.org/TR/WCAG20/#keyboard-operation-keyboard-operable)

**Note:** This issue may be a duplicate with other existing accessibility-related bugs in this project. This issue comes from the Gutenberg accessibility audit, performed by [Tenon](https://www.tenon.io) and funded by [WP Campus](https://wpcampus.org/). This issue is GUT-68 in Tenon's report

**Comment by melchoyce** [comment]
Screenshot from the full report:

![image](https://user-images.githubusercontent.com/2846578/57182858-58a71380-6e59-11e9-9650-2cf984ceaab8.png)

**Comment by afercia** [comment]
This custom component was introduced in #9802. Previously, the font size controls were a set of buttons. Worth revisiting this custom implementation in favor of a native `<select>` element.

**Comment by nicolad** [comment]
@afercia do we need design feedback on this since using `<select>` we will lose that visual cue related to the sizing?
![image](https://user-images.githubusercontent.com/1477453/57712309-03e57300-7679-11e9-95a2-e9f36eb31ba2.png)

**Comment by afercia** [comment]
@nicolad I'd say yes, at least some design feedback is necessary. Personally, I'm not sure the size "preview" in the control is really necessary, as the size change can be immediately seen in the block content.

Regarding the suggested markup, I'd like to discuss it with the accessibility team. As mentioned also in #15307, related form controls don't necessarily need to be always grouped within a `fieldset` + `legend`. Better semantics is always welcome but it's always necessary to take into consideration how markup affects the way controls are actually perceived by assistive technologies.

For example, some screen readers repeat the legend for any control in the fieldset. When the control label already contains the words "font size", I don't see a great practical value to also have a legend that says "font size". That would be just repeated the control gets announced, see screenshot below:

<img width="668" alt="Screenshot 2019-05-14 at 22 46 00" src="https://user-images.githubusercontent.com/1682452/57731064-220f9b00-769a-11e9-93c1-d577df7cf8d3.png">

**Comment by mapk** [comment]
There was a conversation about this in [today's Design Triage in slack](https://wordpress.slack.com/archives/C02S78ZAL/p1558454533496500).

While there was a desire to visually represent the actual font sizes in the dropdown, we understand if it causes issues with assistive devices. It doesn't appear that styling the select box options works across browsers to give the same visual appearance.

Here's a couple thoughts...

**1. Current implementation with added functions**
Is it possible to keep the current implementation and code it so that it works similarly to a native select box?

**2. Native select box with visual icon**
If we convert it to a native select box, maybe we can add a font-sizing icon next to the select box like below. _But after looking at this, I think the icons just muddles it up too much._

![drop](https://user-images.githubusercontent.com/617986/58122052-9deb7500-7bbd-11e9-8bca-b2ced756f055.png)

**3. Native selectbox only**
Just make it a native select box without any visual additions.

![drop-2](https://user-images.githubusercontent.com/617986/58122247-05092980-7bbe-11e9-82c8-f84bb55c7188.png)

**Comment by afercia** [comment]
Discussed during today's accessibility bug-scrub. Agreed to recommend to change this custom control to a native select element.

As per the suggested additional fieldset + legend, sorry I didn't notice the label controls were changed too. That makes sense because the _combined text_ of the legend plus labels would read as:

```
<legend>Font Size</legend>
    Choose basic size
    Custom font size
```

Even better:

```
<legend>Font Size</legend>
    Choose preset
    Custom
```

**Comment by mapk** [comment]
To keep this moving forward, let's change this over to a native select box. 

![native](https://user-images.githubusercontent.com/617986/59232108-f64bdc00-8b97-11e9-9747-49f396949766.png)

We're using native select boxes elsewhere in the UI anyways. Take a look at the Image Size select box.

![nativeselect](https://user-images.githubusercontent.com/617986/59232094-ec29dd80-8b97-11e9-8df9-b7dbfed7af11.gif)

**Comment by drw158** [comment]
As a more long-term approach to this issue, it would be good to add the necessary a11y functions to the custom "select", assuming it is possible and a viable approach. It's a common pattern in many component libraries, and it may be something needed in the future. For example, adding search to the select component for long lists of items. Maybe at that point, it becomes a "dropdown" though, but it still needs to be accessible.

**Comment by afercia** [comment]
>  It's a common pattern in many component libraries

In the last 3 years, I've gone through most of those "custom selects" and none of them is fully accessible. For a simple reason: they implement non-native, non-standard, behaviors and interactions.

The only accessible UI pattern that is close to all these various custom things is the [Combo Box](https://www.w3.org/TR/wai-aria-practices-1.1/#combobox) pattern. It's different though, more an input field with a list of suggestions rather than a custom select. Gutenberg already has a combo box component (the various auto-completers) and it works pretty well.

The point of this issue is to just use a native element rather than adding features to custom components that don't provide full support for browsers APIs and assistive technologies.
